File: Chapter06/HelpDeskCS/HelpDeskCS/HelpDeskCS.DesktopClient/Screens/EditableIssuesGrid.lsml.cs
// You can use and redistribute the code from this book (and sample application) for non-commercial and 
// most commercial purposes as long as you acknowledge the source and authorship. 
// You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
// The acknowledgment should include author, title, publisher, and ISBN. 
// An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2015 by Tim Leung, published by Apress, ISBN 978-1-4842-0767-3".

using Microsoft.LightSwitch.Presentation.Extensions;
using Microsoft.LightSwitch.Presentation;
using Microsoft.LightSwitch.Framework.Client;
using Microsoft.LightSwitch;
using System.Collections.Generic;
using System.IO.IsolatedStorage;
using System.IO;
using System.Linq;
using System;
using System.Text;

namespace LightSwitchApplication
{
    public partial class EditableIssuesGrid
    {

        //Listing 6-8. Discarding screen changes on desktop clients
        partial void EditableIssuesGrid_Saving(ref bool handled)
        {
            StringBuilder message = new StringBuilder();

            foreach (Issue issue in
                    DataWorkspace.ApplicationData.Details.GetChanges().OfType<Issue>())
            {
                message.AppendLine(issue.Id.ToString() + " - " + issue.Subject);

                // You can discard just a single record using this syntax                     
                // issue.Details.DiscardChanges();

                message.AppendLine(issue.Id.ToString());
            }

            if (message.Length > 0)
            {
                if (this.ShowMessageBox("You've made changes to the following records:" +
                        Environment.NewLine +
                        message.ToString() + Environment.NewLine +
                        "Are you sure you wish to save your changes?" +
                        "Click OK to continue, or Cancel to discard your changes",
                        "Confirm Save",
                        MessageBoxOption.OkCancel) == System.Windows.MessageBoxResult.Cancel)
                {
                    DataWorkspace.ApplicationData.Details.DiscardChanges();
                    handled = true;
                }
            }


        }
    }
}

File: Chapter06/HelpDeskVB/HelpDeskVB/HelpDeskVB.DesktopClient/Screens/EditableIssuesGrid.lsml.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2015 by Tim Leung, published by Apress, ISBN 978-1-4842-0767-3".

Imports System.Text

Namespace LightSwitchApplication
    Public Class EditableIssuesGrid
        Private Sub EditableIssuesGrid_SaveError(exception As exception, ByRef handled As Boolean)
            ' Write your code here.

            ' Un-delete deleted records that had server-side validation errors
            Dim validationExc = TryCast(exception, ValidationException)

            If Not validationExc Is Nothing Then
                Dim entities = From v In validationExc.ValidationResults
                               Let e = TryCast(v.Target, IEntityObject)
                               Where Not e Is Nothing AndAlso
      e.Details.EntityState = EntityState.Deleted
                               Select e

                For Each e In entities
                    e.Details.DiscardChanges()
                Next
            End If


        End Sub

        'Listing 6-8. Discarding screen changes on desktop clients
        Private Sub EditableIssuesGrid_Saving(ByRef handled As Boolean)
            ' Write your code here.
            Dim message As StringBuilder = New StringBuilder()
            For Each issue As Issue In
           DataWorkspace.ApplicationData.Details.GetChanges().OfType(Of Issue)()
                message.AppendLine(issue.Id.ToString & " - " & issue.Subject)

                'You can discard just a single record using this syntax                      
                'issue.Details.DiscardChanges()
            Next

            If Me.ShowMessageBox("You've made changes to the following records:" &
        Environment.NewLine & Environment.NewLine &
        message.ToString() & Environment.NewLine &
        "Are you sure you wish to save your changes?" & Environment.NewLine &
        "Click OK to continue, or Cancel to discard your changes",
        "Confirm Save",
         MessageBoxOption.OkCancel) = MessageBoxResult.Cancel Then

                DataWorkspace.ApplicationData.Details.DiscardChanges()
                handled = True
            End If


        End Sub
    End Class
End Namespace
